src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.cpp

```diff
@@ -0,0 +1,893 @@
+        log.warning() << "No decrease clause provided for loop " << function_id
+                      << "." << loop.latch->loop_number << " at "
+                      << loop.head->source_location()
+                      << ". Termination will not be checked." << messaget::eom;
+      }
+    }
+
+    std::vector<std::size_t> inner_loops;
+    for(auto pred_idx : loop_nesting_graph.get_predecessors(loop_id))
+    {
+      inner_loops.push_back(pred_idx);
+    }
+
+    std::vector<std::size_t> outer_loops;
+    for(auto succ_idx : loop_nesting_graph.get_successors(loop_id))
+    {
+      outer_loops.push_back(succ_idx);
+    }
+
+    const auto cbmc_loop_number = loop.latch->loop_number;
+    const auto &language_mode = utils.get_function_symbol(function_id).mode;
+
+    // Generate "write set" variable
+    const auto &write_set_var =
+      get_fresh_aux_symbol(
+        library.dfcc_type[dfcc_typet::WRITE_SET],
+        id2string(function_id),
+        "__write_set_loop_" + std::to_string(cbmc_loop_number),
+        loop.head->source_location(),
+        language_mode,
+        goto_model.symbol_table)
+        .symbol_expr();
+
+    // Generate "address of write set" variable
+    const auto &addr_of_write_set_var =
+      get_fresh_aux_symbol(
+        library.dfcc_type[dfcc_typet::WRITE_SET_PTR],
+        id2string(function_id),
+        "__address_of_write_set_loop_" + std::to_string(cbmc_loop_number),
+        loop.head->source_location(),
+        language_mode,
+        goto_model.symbol_table)
+        .symbol_expr();
+
+    // Insert entry in map and we're done
+    loop_info_map.insert(
+      {loop_id,
+       {loop_id,
+        cbmc_loop_number,
+        assigns,
+        invariant_expr,
+        decreases_expr,
+        loop_local,
+        loop_tracked,
+        inner_loops,
+        outer_loops,
+        write_set_var,
+        addr_of_write_set_var}});
+  }
+};
+
+dfcc_cfg_infot::dfcc_cfg_infot(
+  const irep_idt &function_id,
+  goto_functiont &goto_function,
+  const exprt &top_level_write_set,
+  const dfcc_loop_contract_modet loop_contract_mode,
+  goto_modelt &goto_model,
+  message_handlert &message_handler,
+  dfcc_utilst &utils,
+  dfcc_libraryt &library)
+  : function_id(function_id),
+    goto_function(goto_function),
+    top_level_write_set(top_level_write_set),
+    log(message_handler),
+    ns(goto_model.symbol_table)
+{
+  dfcc_loop_nesting_grapht loop_nesting_graph;
+  goto_programt &goto_program = goto_function.body;
+  messaget log(message_handler);
+  if(loop_contract_mode != dfcc_loop_contract_modet::NONE)
+  {
+    dfcc_check_loop_normal_form(goto_program, log);
+    loop_nesting_graph = build_loop_nesting_graph(goto_program, log);
+    check_inner_loop_have_contracts(loop_nesting_graph);
+
+    auto topsorted = loop_nesting_graph.topsort();
+    for(const auto idx : topsorted)
+    {
+      topsorted_loops.push_back(idx);
+    }
+  }
+
+  tag_loop_instructions(goto_program, loop_nesting_graph);
```

Comment:
Yes, we still want to compute the local, and tracks for top-level, which cover all instructions when `dfcc_loop_contract_modet::NONE`.

Reply:
> So the loop_nesting_graph will be unset in case of dfcc_loop_contract_modet::NONE? Does it make sense to invoke tag_loop_instructions or, really, anything else?

Instruction tags indicate the write set to use for frame condition checks. When loop contracts are not applied the loop graph is empty and all instructions are tagged as top level instructions